apis/kcl/main.k

```diff
@@ -0,0 +1,313 @@
+import regex
+
+xrApiVersion = option("params")?.oxr?.apiVersion
+xrKind = option("params")?.oxr?.kind
+xrName = option("params")?.oxr?.metadata.name
+region = option("params")?.oxr?.spec.parameters.region or ""
+id = option("params")?.oxr?.spec.parameters.id or ""
+vpcCidrBlock = option("params")?.oxr?.spec.parameters.vpcCidrBlock or ""
+subnets = option("params")?.oxr?.spec.parameters.subnets or ""
+providerConfigRefName = option("params")?.oxr?.spec.parameters.providerConfigName or None
+
+_metadata = lambda name = str -> any {
+    {
+        annotations = {"krm.kcl.dev/composition-resource-name" = name}
+    }
+}
+
+_defaults = {
+    deletionPolicy = option("params")?.oxr?.spec.parameters.deletionPolicy or "Delete"
+    if providerConfigRefName:
+        providerConfigRef.name = providerConfigRefName
+}
+
+_cidrEscaped = lambda cidr = str -> str {
+    regex.replace(cidr, "\.|\/", "-")
+}
+
+_formatSubnet = lambda s = dict -> str {
+    "{}-{}-{}".format(s.availabilityZone, _cidrEscaped(s.cidrBlock), s.type)
+}
+
+_getExternalName = lambda resourceName = str -> str {
+    id = option("params")?.ocds?[resourceName]?.Resource?.metadata?.annotations?["crossplane.io/external-name"] or None
+}
+
+_items = [{
+    apiVersion = "ec2.aws.upbound.io/v1beta1"
+    kind = "VPC"
+    metadata = _metadata("vpc") | {
+        labels = {
+            "networks.aws.platform.upbound.io/network-id" = id
+        }
+    }
+    spec = _defaults | {
+        forProvider = {
+            cidrBlock = vpcCidrBlock
+            enableDnsHostnames = True
+            enableDnsSupport = True
+            tags = {
+                Name = xrName
+            }
+            region = region
+        }
+    }
+}]
+
+_items += [{
+    apiVersion = "ec2.aws.upbound.io/v1beta1"
+    kind = "InternetGateway"
+    metadata = _metadata("igw") | {
+        labels = {
+            "networks.aws.platform.upbound.io/network-id" = id
+        }
+    }
+    spec = _defaults | {
+        forProvider = {
+            vpcIdSelector = {
+                matchControllerRef = True
+            }
+            region = region
+        }
+    }
+}]
+
+_items += [{
+    apiVersion = "ec2.aws.upbound.io/v1beta1"
+    kind = "Subnet"
+    metadata = _metadata("subnet-" + _formatSubnet(s)) | {
+        labels = {
+            zone = s.availabilityZone
+            if s.type == "private":
+                access = "private"
+            else:
+                access = "public"
+            "networks.aws.platform.upbound.io/network-id" = id
+        }
+    }
+    spec = _defaults | {
+        forProvider = {
+            cidrBlock = s.cidrBlock
+            if s.type == "public":
```

Comment:
can we cover such conditionals with KCL unit tests, https://github.com/upbound/configuration-azure-network/blob/main/render_test.k style? It might be a great opportunity to introduce them together with the code that is rich with multiple decision points

Reply:
Great idea!

Reply:
tests were added but this needs to be merged before because there was no nice and easy way to run tests for multiple languages: https://github.com/crossplane/build/pull/22/files